runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -585,6 +600,120 @@ function LanguageTree:_iter_regions(fn)
   end
 end
 
+---Whether two region values are approximately equal. Should be implied by equality.
+---See the comment in `set_included_regions` on why we use similarity.
+---For now it simply compares the last bytes of the first and the last regions.
+---@param region1 Range6[]
+---@param region2 Range6[]
+---@return boolean
+local function region_similar(region1, region2)
+  return region1[1][6] == region2[1][6] or region1[#region1][6] == region2[#region2][6]
+end
+
+---@param regions_inv table<integer, (Range6[]|integer)[]>
+---@param region Range6[]
+---@return integer?
+---@return boolean? exact
+local function regions_inv_lookup(regions_inv, region)
+  local bucket = regions_inv[region[1][3]]
+  if not bucket then
+    return
+  end
+
+  local i ---@type integer?
+  for e = 1, #bucket, 2 do
+    local old_region = bucket[e] --[[@as Range6[] ]]
+    if region_similar(old_region, region) then
+      i = bucket[e + 1] --[[@as integer]]
+      if vim.deep_equal(old_region, region) then
+        return i, true
+      end
+    end
+  end
+
+  return i, false
+end
+
+---@param regions_inv table<integer, (Range6[]|integer)[]>
+---@param i integer
+---@param region Range6[]
+local function regions_inv_insert(regions_inv, i, region)
+  local start_byte = region[1][3]
+  local bucket = regions_inv[start_byte]
+  if not bucket then
+    regions_inv[start_byte] = { region, i }
+  else
+    table.insert(bucket, region)
+    table.insert(bucket, i)
+  end
+end
+
+---@param regions_inv table<integer, (Range6[]|integer)[]>
+---@param region Range6[]
+local function regions_inv_remove(regions_inv, region)
```

Comment:
`_discard_region` is called in a loop which calls this function. Each time you call `table.remove` it causes a full table re-index.

Should this be optimized somehow or do you think it won't have much effect?

Reply:
The bucket is very likely to be small, usually contains a single region. So this should be fine.